pkg/component/runtime/manager.go

```diff
@@ -205,87 +216,123 @@ func (m *Manager) Run(ctx context.Context) error {
 			grpc.MaxRecvMsgSize(m.grpcConfig.MaxMsgSize),
 		)
 	}
-	m.netMx.Lock()
-	m.server = server
-	m.netMx.Unlock()
-	proto.RegisterElasticAgentServer(m.server, m)
+	proto.RegisterElasticAgentServer(server, m)
 
 	// start serving GRPC connections
-	var wg sync.WaitGroup
-	wg.Add(1)
+	var wgServer sync.WaitGroup
+	wgServer.Add(1)
 	go func() {
-		defer wg.Done()
-		m.serverReady.Store(true)
-		for {
-			err := server.Serve(lis)
-			if err != nil {
-				m.logger.Errorf("control protocol failed: %s", err)
-			}
-			if ctx.Err() != nil {
-				// context has an error don't start again
-				return
-			}
-		}
+		defer wgServer.Done()
+		go m.serverLoop(ctx, listener, server)
 	}()
 
-	<-ctx.Done()
-	m.running.Store(false)
-	m.shuttingDown.Store(true)
+	// Start the run loop, which continues on the main goroutine
+	// until the context is canceled.
+	m.runLoop(ctx)
+
+	// Notify components to shutdown and wait for their response
 	m.shutdown()
 
+	// Close the rpc listener and wait for serverLoop to return
+	listener.Close()
+	wgServer.Wait()
+
+	// Cancel any remaining connections
 	server.Stop()
-	wg.Wait()
-	m.netMx.Lock()
-	m.listener = nil
-	m.server = nil
-	m.netMx.Unlock()
 	return ctx.Err()
 }
 
+// The main run loop for the runtime manager, whose responsibilities are:
+//   - Accept component model updates from the Coordinator
+//   - Apply those updates safely without ever blocking, because a block here
+//     propagates to a block in the Coordinator
+//   - Close doneChan when the loop ends, so the Coordinator knows not to send
+//     any more updates
+func (m *Manager) runLoop(ctx context.Context) {
+LOOP:
+	for ctx.Err() == nil {
+		select {
+		case <-ctx.Done():
+			break LOOP
+		case model := <-m.updateChan:
+			// We got a new component model from m.Update(), mark it as the
+			// next update to apply, overwriting any previous pending value.
+			m.nextUpdate = &model
+		case <-m.updateDoneChan:
+			// An update call has finished, we can initiate another when available.
+			m.updateInProgress = false
+		}
+
+		// After each select call, check if there's a pending update that
+		// can be applied.
+		if m.nextUpdate != nil && !m.updateInProgress {
+			// There is a component model update available, apply it.
+			go func(model component.Model) {
+				// Run the update with tearDown set to true since this is coming
+				// from a user-initiated policy update
+				result := m.update(model, true)
+
+				// When update is done, send its result back to the coordinator,
+				// unless we're shutting down.
+				select {
+				case m.errCh <- result:
```

Comment:
Nit: the result here is an error.
```suggestion
				err := m.update(model, true)

				// When update is done, send its result back to the coordinator,
				// unless we're shutting down.
				select {
				case m.errCh <- err:
```